src/framegrab/rtsp_discovery.py

```diff
@@ -0,0 +1,150 @@
+import logging
+import urllib.parse
+from typing import List, Optional
+
+import onvif
+from onvif import ONVIFCamera
+from pydantic import BaseModel
+from wsdiscovery import QName
+from wsdiscovery.discovery import ThreadedWSDiscovery as WSDiscovery
+
+logger = logging.getLogger(__name__)
+
+
+# Default credentials to try when connecting to RTSP cameras, used in discover_camera_ips() when try_default_logins=True
+DEFAULT_CREDENTIALS = [
+    ("admin", "admin"),
+    ("admin", ""),
+    ("", ""),
+    ("root", "camera"),
+    ("root", "root"),
+    ("admin", "12345"),
+    ("admin", "123456"),
+    ("admin", "password"),
+    ("user", "user"),
+    ("root", "pass"),
+]
+
+
+# Model for storing ONVIF Device Information
+class ONVIFDeviceInfo(BaseModel):
+    ip: str
+    port: Optional[int] = 80
+    username: Optional[str] = ""
+    password: Optional[str] = ""
+    xaddr: Optional[str] = ""  # ONVIF service address
+    rtsp_urls: Optional[List[str]] = []  # List of rtsp urls for the camera
+
+
+class RTSPDiscovery:
+    # Simple RTSP camera discovery with ONVIF capabilities
+
+    @staticmethod
+    def discover_camera_ips(try_default_logins: bool = False) -> List[ONVIFDeviceInfo]:
+        """
+        Uses WSDiscovery to find ONVIF supported devices.
+
+        Parameters:
+        try_default_logins (bool, optional): Option to try different default credentials stored in DEFAULT_CREDENTIALS.
+            Defaults to False.
+
+        Returns:
+        List[ONVIFDeviceInfo]: A list of ONVIFDeviceInfos with IP address, port number, and ONVIF service address.
+        """
+
+        device_ips = []
+        logger.debug("Starting WSDiscovery for ONVIF devices")
+        wsd = WSDiscovery()
+        wsd.start()
+        types = [QName("http://www.onvif.org/ver10/network/wsdl", "NetworkVideoTransmitter")]
+        ret = wsd.searchServices(types=types)
+        for service in ret:
+            xaddr = service.getXAddrs()[0]
+            parsed_url = urllib.parse.urlparse(xaddr)
+            ip = parsed_url.hostname
+            port = parsed_url.port or 80  # Use the default port 80 if not specified
+
+            logger.debug(f"Found ONVIF service at {xaddr}")
+            device_ip = ONVIFDeviceInfo(ip=ip, port=port, username="", password="", xaddr=xaddr, rtsp_urls=[])
+
+            if try_default_logins:
+                RTSPDiscovery._try_logins(device=device_ip)
+
+            device_ips.append(device_ip)
+        wsd.stop()
+        return list(device_ips)
+
+    @staticmethod
+    def generate_rtsp_urls(device: ONVIFDeviceInfo) -> bool:
```

Comment:
I would expect a function with this name to return rtsp_urls

Reply:
Yea should be just returning an empty list when wrong credentials, will update the function to be more in line with its name.